File: src/main/java/no/fint/provider/springer/service/IdentifikatorFactory.java
package no.fint.provider.springer.service;

import lombok.extern.slf4j.Slf4j;
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.provider.springer.storage.Springer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;
import java.util.concurrent.atomic.AtomicInteger;

@Service
@Slf4j
public class IdentifikatorFactory {

    @Autowired(required = false)
    private MongoTemplate mongoTemplate;

    @PostConstruct
    public void init() {
        if (mongoTemplate != null) {
            try {
                long count = mongoTemplate.count(null, Springer.class);
                atomicInteger.set((int) count);
                log.info("Setting initial value to {}", atomicInteger.get());
            } catch (Exception e) {
                log.info("Unable to update initial value: {} {}", e.getClass().getSimpleName(), e.getMessage());
            }
        }
    }

    private AtomicInteger atomicInteger = new AtomicInteger();

    public Identifikator create() {
        Identifikator identifikator = new Identifikator();
        identifikator.setIdentifikatorverdi("ID%010d".formatted(atomicInteger.incrementAndGet()));
        return identifikator;
    }
}

File: src/main/kotlin/no/fint/provider/springer/service/IdentifikatorFactory.kt
package no.fint.provider.springer.service

import jakarta.annotation.PostConstruct
import no.fint.provider.springer.storage.Springer
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator
import org.slf4j.LoggerFactory
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.query.Query
import org.springframework.stereotype.Service
import java.util.concurrent.atomic.AtomicInteger

@Service
class IdentifikatorFactory(
    private val mongoTemplate: MongoTemplate?
) {
    private val log = LoggerFactory.getLogger(IdentifikatorFactory::class.java)
    private val atomicInteger = AtomicInteger()

    @PostConstruct
    fun init() {
        mongoTemplate?.let {
            try {
                val count = it.count(Query(), Springer::class.java)
                atomicInteger.set(count.toInt())
                log.info("Setting initial value to {}", atomicInteger.get())
            } catch (e: Exception) {
                log.info("Unable to update initial value: {} {}", e.javaClass.simpleName, e.message)
            }
        }
    }

    fun create(): Identifikator = Identifikator().apply {
        identifikatorverdi = "ID%010d".format(atomicInteger.incrementAndGet())
    }
}
